[Bug]: Indexcrood error

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:
- Deployment mode(standalone or cluster):
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```


### Current Behavior

While checking Milvus pods, indexcrood got errors:
[2023/03/21 11:28:38.282 +00:00] [ERROR] [indexcoord/meta_table.go:149] ["failed to alter segments index in meta store"] ["segment indexes num"=152] [error="etcdserver: too many operations in txn request"] [stack="github.com/milvus-io/milvus/internal/indexcoord.(*metaTable).alterSegmentIndexes\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/meta_table.go:149\ngithub.com/milvus-io/milvus/internal/indexcoord.(*metaTable).MarkSegmentsIndexAsDeleted\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/meta_table.go:794\ngithub.com/milvus-io/milvus/internal/indexcoord.(*IndexCoord).DropIndex\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/index_coord.go:769\ngithub.com/milvus-io/milvus/internal/distributed/indexcoord.(*Server).DropIndex\n\t/go/src/github.com/milvus-io/milvus/internal/distributed/indexcoord/service.go:283\ngithub.com/milvus-io/milvus/internal/proto/indexpb._IndexCoord_DropIndex_Handler.func1\n\t/go/src/github.com/milvus-io/milvus/internal/proto/indexpb/index_coord.pb.go:2513\ngithub.com/milvus-io/milvus/internal/util/logutil.UnaryTraceLoggerInterceptor\n\t/go/src/github.com/milvus-io/milvus/internal/util/logutil/grpc_interceptor.go:22\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25\ngithub.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.UnaryServerInterceptor.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/tracing/opentracing/server_interceptors.go:38\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34\ngithub.com/milvus-io/milvus/internal/proto/indexpb._IndexCoord_DropIndex_Handler\n\t/go/src/github.com/milvus-io/milvus/internal/proto/indexpb/index_coord.pb.go:2515\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:1283\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:1620\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.2\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:922"]


[2023/03/21 11:28:38.282 +00:00] [ERROR] [indexcoord/index_coord.go:778] ["IndexCoord drop index fail"] [collectionID=439941373613007856] [partitionIDs="[439941373676815182]"] [indexName=] [error="etcdserver: too many operations in txn request"] [stack="github.com/milvus-io/milvus/internal/indexcoord.(*IndexCoord).DropIndex\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/index_coord.go:778\ngithub.com/milvus-io/milvus/internal/distributed/indexcoord.(*Server).DropIndex\n\t/go/src/github.com/milvus-io/milvus/internal/distributed/indexcoord/service.go:283\ngithub.com/milvus-io/milvus/internal/proto/indexpb._IndexCoord_DropIndex_Handler.func1\n\t/go/src/github.com/milvus-io/milvus/internal/proto/indexpb/index_coord.pb.go:2513\ngithub.com/milvus-io/milvus/internal/util/logutil.UnaryTraceLoggerInterceptor\n\t/go/src/github.com/milvus-io/milvus/internal/util/logutil/grpc_interceptor.go:22\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25\ngithub.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.UnaryServerInterceptor.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/tracing/opentracing/server_interceptors.go:38\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34\ngithub.com/milvus-io/milvus/internal/proto/indexpb._IndexCoord_DropIndex_Handler\n\t/go/src/github.com/milvus-io/milvus/internal/proto/indexpb/index_coord.pb.go:2515\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:1283\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:1620\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.2\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:922"]

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Milvus Log

_No response_

### Anything else?

_No response_

occurred in  milvus:v2.2.8   
docker image (milvusdb/milvus:v2.2.8)

[2023/07/14 01:29:24.771 +00:00] [ERROR] [indexcoord/meta_table.go:149] ["failed to alter segments index in meta store"] ["segment indexes num"=631] [error="etcdserver: too many operations in txn request"] [stack="github.com/milvus-io/milvus/internal/indexcoord.(*metaTable).alterSegmentIndexes\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/meta_table.go:149\ngithub.com/milvus-io/milvus/internal/indexcoord.(*metaTable).MarkSegmentsIndexAsDeleted\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/meta_table.go:801\ngithub.com/milvus-io/milvus/internal/indexcoord.(*IndexCoord).DropIndex\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/index_coord.go:761\ngithub.com/milvus-io/milvus/internal/distributed/indexcoord.(*Server).DropIndex\n\t/go/src/github.com/milvus-io/milvus/internal/distributed/indexcoord/service.go:283\ngithub.com/milvus-io/milvus/internal/proto/indexpb._IndexCoord_DropIndex_Handler.func1\n\t/go/src/github.com/milvus-io/milvus/internal/proto/indexpb/index_coord.pb.go:2513\ngithub.com/milvus-io/milvus/internal/util/logutil.UnaryTraceLoggerInterceptor\n\t/go/src/github.com/milvus-io/milvus/internal/util/logutil/grpc_interceptor.go:22\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25\ngithub.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.UnaryServerInterceptor.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/tracing/opentracing/server_interceptors.go:38\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34\ngithub.com/milvus-io/milvus/internal/proto/indexpb._IndexCoord_DropIndex_Handler\n\t/go/src/github.com/milvus-io/milvus/internal/proto/indexpb/index_coord.pb.go:2515\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:1283\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:1620\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.2\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:922"]
[2023/07/14 01:29:24.771 +00:00] [ERROR] [indexcoord/index_coord.go:770] ["IndexCoord drop index fail"] [collectionID=442660375745463817] [partitionIDs="[442660375749549807]"] [indexName=] [error="etcdserver: too many operations in txn request"] [stack="github.com/milvus-io/milvus/internal/indexcoord.(*IndexCoord).DropIndex\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/index_coord.go:770\ngithub.com/milvus-io/milvus/internal/distributed/indexcoord.(*Server).DropIndex\n\t/go/src/github.com/milvus-io/milvus/internal/distributed/indexcoord/service.go:283\ngithub.com/milvus-io/milvus/internal/proto/indexpb._IndexCoord_DropIndex_Handler.func1\n\t/go/src/github.com/milvus-io/milvus/internal/proto/indexpb/index_coord.pb.go:2513\ngithub.com/milvus-io/milvus/internal/util/logutil.UnaryTraceLoggerInterceptor\n\t/go/src/github.com/milvus-io/milvus/internal/util/logutil/grpc_interceptor.go:22\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25\ngithub.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.UnaryServerInterceptor.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/tracing/opentracing/server_interceptors.go:38\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34\ngithub.com/milvus-io/milvus/internal/proto/indexpb._IndexCoord_DropIndex_Handler\n\t/go/src/github.com/milvus-io/milvus/internal/proto/indexpb/index_coord.pb.go:2515\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:1283\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:1620\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.2\n\t/go/pkg/mod/google.golang.org/grpc@v1.46.0/server.go:922"]
[2023/07/14 01:29:24.772 +00:00] [WARN] [retry/retry.go:44] ["retry func failed"] ["retry time"=0] [error="etcdserver: too many operations in txn request"]
[2023/07/14 01:29:24.772 +00:00] [WARN] [rootcoord/step_executor.go:74] ["failed to execute step, wait for reschedule"] [error="etcdserver: too many operations in txn request"] [step="drop collection index: 442660375745463817"]

--- a/internal/metastore/kv/indexcoord/kv_catalog.go
+++ b/internal/metastore/kv/indexcoord/kv_catalog.go
@@ -149,8 +149,19 @@ func (kc *Catalog) AlterSegmentIndexes(ctx context.Context, segIdxes []*model.Se
 			return err
 		}
 		kvs[key] = string(value)
+		// TODO when we have better txn kv we should make this as a transaction
+		if len(kvs) >= 64 {
+			err := kc.Txn.MultiSave(kvs)
+			if err != nil {
+				return err
+			}
+			kvs = make(map[string]string)
+		}
 	}
-	return kc.Txn.MultiSave(kvs)
+	if len(kvs) != 0 {
+		return kc.Txn.MultiSave(kvs)
+	}
+	return nil
 }
 
 func (kc *Catalog) DropSegmentIndex(ctx context.Context, collID, partID, segID, buildID typeutil.UniqueID) error {

--- a/internal/metastore/kv/indexcoord/kv_catalog_test.go
+++ b/internal/metastore/kv/indexcoord/kv_catalog_test.go
@@ -384,6 +384,45 @@ func TestCatalog_AlterSegmentIndexes(t *testing.T) {
 	})
 }
 
+func TestCatalog_AlterSegmentMultiIndexes(t *testing.T) {
+	segIdx := make([]*model.SegmentIndex, 100)
+	for i := 0; i < 100; i++ {
+		segIdx[i] = &model.SegmentIndex{
+			SegmentID:     int64(i),
+			CollectionID:  0,
+			PartitionID:   0,
+			NumRows:       0,
+			IndexID:       0,
+			BuildID:       0,
+			NodeID:        0,
+			IndexState:    0,
+			FailReason:    "",
+			IndexVersion:  0,
+			IsDeleted:     false,
+			CreateTime:    0,
+			IndexFileKeys: nil,
+			IndexSize:     0,
+		}
+	}
+
+	t.Run("batchAdd", func(t *testing.T) {
+		txn := &MockedTxnKV{
+			multiSave: func(kvs map[string]string) error {
+				if len(kvs) >= 128 {
+					return errors.New("txn too large")
+				}
+				return nil
+			},
+		}
+		catalog := &Catalog{
+			Txn: txn,
+		}
+
+		err := catalog.AlterSegmentIndexes(context.Background(), segIdx)
+		assert.NoError(t, err)
+	})
+}
+
 func TestCatalog_DropSegmentIndex(t *testing.T) {
 	t.Run("success", func(t *testing.T) {
 		txn := &MockedTxnKV{

